python.d/nvidia_smi.chart.py

```diff
@@ -0,0 +1,255 @@
+        self.break_condition = break_condition
+        self.lock = threading.RLock()
+        self.last_time = 0
+        self.last_data = ''
+
+    def _read(self):
+        lines = []
+        while True:
+            line = self.process.stdout.readline().decode()
+            lines.append(line)
+            if self.break_condition(line):
+                self.lock.acquire()
+                self.last_data = '\n'.join(lines)
+                self.last_time = time.time()
+                self.lock.release()
+                return
+
+    def run(self):
+        self.lock.acquire()
+        try:
+            self.process = subprocess.Popen(self.command, stdout=subprocess.PIPE)
+        except (OSError, FileNotFoundError):
+            return
+
+        # Initial read
+        self._read()
+        self.lock.release()
+
+        while True:
+            self._read()
+
+
+class Service(SimpleService):
+    def __init__(self, configuration=None, name=None):
+        SimpleService.__init__(self, configuration=configuration, name=name)
+        self.order = []
+        self.definitions = {}
+        self.fake_name = "gpu"
+        self.nvidia_smi = find_binary('nvidia-smi')
+        self.assignment = {}
+        self.poller = Poller([self.nvidia_smi, '-x', '-q', '-l', '1'], lambda x: '</nvidia_smi_log>' in x)
+
+    def _invoke_nvidia_smi(self):
+        if self.poller.ident is None:
+            proc = subprocess.Popen([self.nvidia_smi, '-x', '-q'], stdout=subprocess.PIPE)
+            stdout, _ = proc.communicate()
+        else:
+            # TODO: Warn if self.poller.last_time is old.
+            self.poller.lock.acquire()
+            stdout = self.poller.last_data
+            self.poller.lock.release()
+        try:
+            smi = et.fromstring(stdout)
+        except et.ParseError:
+            return None
+        return smi
+
+    def _get_data(self):
+        data = {}
+
+        smi = self._invoke_nvidia_smi()
+
+        if smi is None:
+            return data
+
+        for gpu in smi.findall('gpu'):
+            gpuid = self.assignment[gpu.get('id')]['gpuid']
+
+            for template in CHART_TEMPLATES.values():
+                for rootname, metrics in template['_metrics'].items():
+                    if rootname:
+                        root = gpu.find(rootname)
+                    else:
+                        root = gpu
+
+                    for metric in metrics:
+                        metric_name = gpuid + "_" + metric
+                        elem = root.find(metric)
+
+                        if elem is None:
+                            continue
+
+                        try:
+                            val = float(elem.text.split()[0])
+                        except ValueError:
+                            continue
+
+                        val *= float(template.get('_divisor', 1))
+
+                        data[metric_name] = int(val)
+
+        return data
+
+    def check(self):
```

Comment:
`check` method is super long and hard to read. Please split it into several functions. At least let's move chart creating to a separate function.